dbms/src/Flash/Coprocessor/JoinInterpreterHelper.cpp

```diff
@@ -0,0 +1,332 @@
+    }
+
+    return {chain.getLastActions(), std::move(filter_column_for_other_condition), std::move(filter_column_for_other_eq_condition)};
+}
+} // namespace
+
+TiflashJoin::TiflashJoin(const tipb::Join & join_)
+    : join(join_)
+    , join_key_types(getJoinKeyTypes(join_))
+    , join_key_collators(getJoinKeyCollators(join_, join_key_types))
+{
+    std::tie(kind, build_side_index) = getJoinKindAndBuildSideIndex(join);
+    strictness = isSemiJoin() ? ASTTableJoin::Strictness::Any : ASTTableJoin::Strictness::All;
+}
+
+String TiflashJoin::genMatchHelperName(const Block & header1, const Block & header2) const
+{
+    if (!isLeftSemiFamily())
+    {
+        return "";
+    }
+
+    size_t i = 0;
+    String match_helper_name = fmt::format("{}{}", Join::match_helper_prefix, i);
+    while (header1.has(match_helper_name) || header2.has(match_helper_name))
+    {
+        match_helper_name = fmt::format("{}{}", Join::match_helper_prefix, ++i);
+    }
+    return match_helper_name;
+}
+
+NamesAndTypes TiflashJoin::genColumnsForOtherJoinFilter(
+    const Block & left_input_header,
+    const Block & right_input_header,
+    const ExpressionActionsPtr & prepare_join_actions1,
+    const ExpressionActionsPtr & prepare_join_actions2) const
+{
+    /// columns_for_other_join_filter is a vector of columns used
+    /// as the input columns when compiling other join filter.
+    /// Note the order in the column vector is very important:
+    /// first the columns in left_input_header, then followed
+    /// by the columns in right_input_header, if there are other
+    /// columns generated before compile other join filter, then
+    /// append the extra columns afterwards. In order to figure out
+    /// whether a given column is already in the column vector or
+    /// not quickly, we use another set to store the column names
+
+    NamesAndTypes columns_for_other_join_filter;
+    std::unordered_set<String> column_set_for_other_join_filter;
+
+    auto append_columns = [&columns_for_other_join_filter, &column_set_for_other_join_filter](const Block & header, bool make_nullable) {
+        for (auto const & p : header)
+        {
+            columns_for_other_join_filter.emplace_back(p.name, make_nullable ? makeNullable(p.type) : p.type);
+            column_set_for_other_join_filter.emplace(p.name);
+        }
+    };
+    append_columns(left_input_header, join.join_type() == tipb::JoinType::TypeRightOuterJoin);
+    append_columns(right_input_header, join.join_type() == tipb::JoinType::TypeLeftOuterJoin);
+
+    auto append_unique_columns = [&columns_for_other_join_filter, &column_set_for_other_join_filter](const Block & header) {
+        for (auto const & p : header)
+        {
+            if (column_set_for_other_join_filter.find(p.name) == column_set_for_other_join_filter.end())
+                columns_for_other_join_filter.emplace_back(p.name, p.type);
```

Comment:
I believe we should make the type of columns from the outer side nullable for outer join. It should be a bug in the original implementation.

Reply:
> Don't need to update the `column_set_for_other_join_filter`?

> Don't need to update the `column_set_for_other_join_filter`?

I don't think it is necessary.
Because the new columns in prepareJoin are filter column and join_keys, it is impossible to repeat.

Reply:
> I believe we should make the type of columns from the outer side nullable for outer join. It should be a bug in the original implementation.

The new columns in prepareJoin are filter column and join_keys, not the output column of join.

Reply:
> Don't need to update the `column_set_for_other_join_filter`?

The new columns in prepareJoin are filter column and join_keys.
It is impossible that `filter_column_of_left == filter_column_of_right` and `join_keys_of_left.contains_any_of(join_keys_of_right)`.

Reply:
> > I believe we should make the type of columns from the outer side nullable for outer join. It should be a bug in the original implementation.
> 
> The new columns in prepareJoin are filter column and join_keys, not the output column of join. The reason to make the type of columns from the outer side nullable for outer join is because outer join will fill null columns when they don't match, but this doesn't happen with join keys and filter column

1. There is a typo in my previous comments, it should be "make the type of columns from the **inner** side nullable for outer join"
2. The columns where `join_other_condition` used are columns after join, so the column from inner side of outer join is always nullable.
3. There is another bug here: the extra column from build side should not be added here, because in `join`, only `ColumnsAddedByJoin` are added, so no extra columns from build side will be added to the result of join.